File: main/binary-tree-preorder-traversal/binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    static vector<int> preorderTraversal(const TreeNode* root);
};

vector<int> Solution::preorderTraversal(const TreeNode* root)
{
    vector<int> ret;
    
    stack<const TreeNode*> s;
    s.push(root);
    
    while (!s.empty()) {
        root = s.top();
        s.pop();
        
        if (!root) continue;
        
        ret.push_back(root->val);
        
        s.push(root->right);
        s.push(root->left);
    }
    
    return ret;
}

File: main/binary-tree-preorder-traversal/binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        final List<Integer> vals = new ArrayList<>();
        final Deque<TreeNode> stack = new ArrayDeque<>();

        while (root != null || !stack.isEmpty()) {
            // Go left as far as possible.
            for (; root != null; root = root.left) {
                vals.add(root.val);
                stack.push(root);
            }

            // Go right.
            root = stack.pop().right;
        }

        return vals;
    }
}
